posts/simprocs-intro.md

```diff
@@ -0,0 +1,550 @@
+
+To see what steps `simp` performs, one can use `set_option trace.Meta.Tactic.simp true`, which prints some of this information in the infoview.
+
+Here is an example of a series of rewrites performed by `simp` on a simple goal:
+```lean
+set_option trace.Meta.Tactic.simp true in
+example : 37 * (Nat.fib 0 + 0) = 0 := by
+  simp
+```
+which yields
+```
+[Meta.Tactic.simp.rewrite] Nat.fib_zero:1000:
+      Nat.fib 0
+    ==>
+      0
+[Meta.Tactic.simp.rewrite] add_zero:1000:
+      0 + 0
+    ==>
+      0
+[Meta.Tactic.simp.rewrite] mul_zero:1000:
+      37 * 0
+    ==>
+      0
+[Meta.Tactic.simp.rewrite] eq_self:1000:
+      0 = 0
+    ==>
+      True
+```
+i.e. the steps are:
+```lean
+⊢ 37 * (Nat.fib 0 + 0) = 0
+⊢ 37 * (0 + 0) = 0
+⊢ 37 * 0 = 0
+⊢ 0 = 0
+⊢ True
+```
+
+In this picture, simp lemmas are *fixed* rules to turn a *specific* left hand side into a *specific* right hand side.
+In contrast, simprocs are *flexible* rules to turn a *specific* left hand side into a right hand side *computed* from the left hand side.
+In this sense, they are *parametric simp lemmas*.
+
+## Examples of simprocs
+
+In this section, we exemplify three simprocs through the following use cases:
+* Avoiding combinatorial explosion of lemmas
+* Performance optimisation
+* Computation
+
+### Avoiding combinatorial explosion of lemmas: The `existsAndEq` simproc
+
+The `existsAndEq` simproc is designed to simplify expressions of the form `∃ a, ... ∧ a = a' ∧ ...` where `a'` is some quantity independent of `a'` by removing the existential quantifier and replacing all occurences of `a` by `a'`.
+
+```lean
+example : ∃ (a : ℤ), a*a = 25 ∧ a = 5 := by
+  simp +arith only [existsAndEq, and_true]
+
+example : (∃ a, (∃ b, a + b = 5) ∧ a = 3) ↔ ∃ b, 3 + b = 5 := by
+  simp only [existsAndEq, and_true]
+```
+
+Roughly speaking, the way this metaprogram operates is a follows: whenever an expression of the form `∃ a, P a` with `P a = Q ∧ R`  is encountered:
+- Recursively traverse the expression of the predicate inside the existential quantifier to try and detect an equality `a = a'` by splitting any `∧` that is found along the way into its components.
+- If an equality is found, construct a proof that `P a` implies `a = a`.
+- Construct a proof that `(∃ a, P a) = P a'` using the following theorem
+```lean
+theorem exists_of_imp_eq {α : Sort u} {p : α → Prop} (a : α) (h : ∀ b, p b → a = b) :
+    (∃ b, p b) = p a
+```
+
+### Performance optimisation: The `reduceIte` simproc
+
+The `reduceIte` simproc is designed to take expressions of the form `ite P a b` and replace them with `a` or `b`, depending on whether `P` can be simplified to `True` or `False` by a `simp` call.
+
+```lean
+example : ite (1 + 1 = 2) 1 2 = 1 := by
+  -- Works since `simp` can simplify `1 + 1 = 2` to `True`.
```

Comment:
This might be true today but it's not really `simp`s job to do this. Why not choose a better example like `0 * 37 = 0`, something which is clearly `simp`s job even for nonexperts? Maybe also explain that `simp` can do this because `zero_mul` is a simp lemma?